src/net/data_stream.rs

```diff
@@ -73,216 +107,143 @@ pub struct AdcDacData {
 
-    let generator = BlockGenerator::new(producer);
+    let frame_pool =
+        cortex_m::singleton!(: Pool<[u8; FRAME_SIZE]>= Pool::new()).unwrap();
+
+    // Note(unsafe): We guarantee that FRAME_DATA is only accessed once in this function.
+    let memory = unsafe { &mut FRAME_DATA };
+    frame_pool.grow(memory);
+
+    let generator = FrameGenerator::new(producer, frame_pool);
 
-    let stream = DataStream::new(stack, consumer);
+    let stream = DataStream::new(stack, consumer, frame_pool);
 
     (generator, stream)
 }
 
-/// The data generator for a stream.
-pub struct BlockGenerator {
-    queue: Producer<'static, AdcDacData, BLOCK_BUFFER_SIZE>,
-    current_id: u16,
+struct StreamFrame {
+    format: StreamFormat,
+    sequence_number: u16,
+    buffer: Box<[u8; FRAME_SIZE], Init>,
+    offset: usize,
+    batch_count: u16,
+    batch_size: u8,
 }
 
-impl BlockGenerator {
-    /// Construct a new generator.
-    /// # Args
-    /// * `queue` - The producer portion of the SPSC queue to enqueue data into.
-    ///
-    /// # Returns
-    /// The generator to use.
-    fn new(queue: Producer<'static, AdcDacData, BLOCK_BUFFER_SIZE>) -> Self {
+impl StreamFrame {
+    pub fn new(
+        buffer: Box<[u8; FRAME_SIZE], Uninit>,
+        format: StreamFormat,
+        sequence_number: u16,
+    ) -> Self {
         Self {
-            queue,
-            current_id: 0,
+            format,
+            offset: 7,
+            sequence_number,
+            buffer: unsafe { buffer.assume_init() },
+            batch_size: SAMPLE_BUFFER_SIZE as u8,
+            batch_count: 0,
         }
     }
 
-    /// Schedule data to be sent by the generator.
-    ///
-    /// # Note
-    /// If no space is available, the data batch may be silently dropped.
-    ///
-    /// # Args
-    /// * `adcs` - The ADC data to transmit.
-    /// * `dacs` - The DAC data to transmit.
-    pub fn send(
-        &mut self,
-        adcs: &[&mut [u16; SAMPLE_BUFFER_SIZE]; 2],
-        dacs: &[&mut [u16; SAMPLE_BUFFER_SIZE]; 2],
-    ) {
-        let block = AdcDacData {
-            block_id: self.current_id,
-            adcs: [*adcs[0], *adcs[1]],
-            dacs: [*dacs[0], *dacs[1]],
-        };
-
-        self.current_id = self.current_id.wrapping_add(1);
-        self.queue.enqueue(block).ok();
+    pub fn add_batch<F, const T: usize>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut [u8]),
+    {
+        assert!(!self.is_full::<T>(), "Batch cannot be added to full frame");
+
+        let result = f(&mut self.buffer[self.offset..self.offset + T]);
+
+        self.offset += T;
+        self.batch_count = self.batch_count.checked_add(1).unwrap();
+
+        result
+    }
+
+    pub fn is_full<const T: usize>(&self) -> bool {
+        self.offset + T >= self.buffer.len()
+    }
+
+    pub fn finish(&mut self) -> &[u8] {
+        let offset = self.offset;
+        self.buffer[0..2].copy_from_slice(&self.sequence_number.to_ne_bytes());
```

Comment:
Ah. `ne` is not "network" but "native"...

Reply:
This threw me through a loop as well.